framework/src/userobject/ShapeElementUserObject.C

```diff
@@ -0,0 +1,32 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ShapeElementUserObject.h"
+
+template<>
+InputParameters validParams<ShapeElementUserObject>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  return params;
+}
+
+ShapeElementUserObject::ShapeElementUserObject(const InputParameters & parameters) :
+    ElementUserObject(parameters),
+    _phi(_assembly.phi()),
```

Comment:
Ok, got it. Of course the ```_j``` indexes only the local dofs of _one_ variable, not all local DOFs on the element. In ```ComputeUserObjectsThread``` I loop over all variables and prepare the phis, but what that does is just overwrite them a bunch of times.

We'd need an ```executeJacobian(unsigne int jvar)``` method that gets called after a variable is prepared. I still maintain the point that Jacobian storage should be up to the user (just like the interface for the ```residual``` data of teh UserObject).